pkg/cloud/metadata_test.go

```diff
@@ -267,143 +218,123 @@ func TestNewMetadataService(t *testing.T) {
 				InstanceID:       stdInstanceID,
 				InstanceType:     stdInstanceType,
 				Region:           stdRegion,
 				AvailabilityZone: stdAvailabilityZone,
 			},
-			getInstanceDocErr: nil,
-			getOutpostArnErr:  fmt.Errorf("405"),
+			getMetadataError: fmt.Errorf("405"),
+			expectedErr:      fmt.Errorf("something went wrong while getting EC2 outpost arn: 405"),
 		},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
 			clientset := fake.NewSimpleClientset(&tc.node)
-			if tc.name == "failure: metadata not available, k8s client error" {
-				clientset.PrependReactor("get", "*", func(action k8s_testing.Action) (handled bool, ret runtime.Object, err error) {
-					return true, nil, fmt.Errorf("client failure")
-				})
+			clientsetInitialized := false
+			if tc.clientsetReactors != nil {
+				tc.clientsetReactors(clientset)
 			}
+
 			mockCtrl := gomock.NewController(t)
 			mockEC2Metadata := mocks.NewMockEC2Metadata(mockCtrl)
 
-			mockEC2Metadata.EXPECT().Available().Return(tc.isAvailable)
-			os.Setenv("CSI_NODE_NAME", tc.nodeNameEnvVar)
-			if tc.isAvailable {
-				mockEC2Metadata.EXPECT().GetInstanceIdentityDocument().Return(tc.identityDocument, tc.getInstanceDocErr)
-			}
+			ec2MetadataClient := func() (EC2Metadata, error) { return mockEC2Metadata, nil }
+			k8sAPIClient := func() (kubernetes.Interface, error) { clientsetInitialized = true; return clientset, nil }
 
-			if tc.isAvailable && tc.getInstanceDocErr == nil && !tc.isPartial {
-				mockEC2Metadata.EXPECT().GetMetadata(OutpostArnEndpoint).Return(tc.rawOutpostArn, tc.getOutpostArnErr)
-			}
+			mockEC2Metadata.EXPECT().Available().Return(tc.ec2metadataAvailable)
+			if tc.ec2metadataAvailable {
+				mockEC2Metadata.EXPECT().GetInstanceIdentityDocument().Return(tc.getInstanceIdentityDocumentValue, tc.getInstanceIdentityDocumentError)
 
-			m, err := NewMetadataService(mockEC2Metadata, clientset)
-			if tc.isAvailable && tc.getInstanceDocErr == nil && tc.getOutpostArnErr == nil && !tc.isPartial {
-				if err != nil {
-					t.Fatalf("NewMetadataService() failed: expected no error, got %v", err)
+				// GetMetadata is to get the outpost ARN. It should be skipped if
+				// GetInstanceIdentityDocument returns an error or (somehow?) partial
+				// output
+				if tc.getInstanceIdentityDocumentError == nil && !tc.invalidInstanceIdentityDocument {
+					if tc.getMetadataValue != "" || tc.getMetadataError != nil {
+						mockEC2Metadata.EXPECT().GetMetadata(OutpostArnEndpoint).Return(tc.getMetadataValue, tc.getMetadataError)
+					} else {
+						mockEC2Metadata.EXPECT().GetMetadata(OutpostArnEndpoint).Return("", fmt.Errorf("404"))
+					}
 				}
-
-				if m.GetInstanceID() != tc.identityDocument.InstanceID {
-					t.Fatalf("GetInstanceID() failed: expected %v, got %v", tc.identityDocument.InstanceID, m.GetInstanceID())
+				if clientsetInitialized == true {
+					t.Errorf("kubernetes client was unexpectedly initialized when metadata is available!")
```

Comment:
here is the relevant part where we test that clientset initialization func did not get called.

rest of changes in this test file are mainly renaming and simplifying/collapsing some branching